designs/2021-schema-object-rules/README.md

```diff
@@ -0,0 +1,274 @@
+
+### Documentation for requiring object-style rules
+
+We can finally delete `docs/developer-guide/working-with-rules-deprecated.md`, along with updating/removing any other documentation or test cases that use function-style rules.
+
+## Drawbacks
+
+<!--
+    Why should we *not* do this? Consider why adding this into ESLint
+    might not benefit the project or the community. Attempt to think
+    about any opposing viewpoints that reviewers might bring up.
+
+    Any change has potential downsides, including increased maintenance
+    burden, incompatibility with other tools, breaking existing user
+    experience, etc. Try to identify as many potential problems with
+    implementing this RFC as possible.
+-->
+
+### Drawbacks of requiring schemas
+
+Plugin authors who left out schemas will need to take the time to add them.
+
+Users of old/unmaintained plugins/rules which are missing schemas will not be able to use them until they are updated.
+
+### Drawbacks of requiring object-style rules
+
+Plugin authors who still have function-style rules will need to convert them to object-style rules.
+
+Users of old/unmaintained plugins/rules which still have function-style rules will not be able to use them until they are updated.
+
+## Backwards Compatibility Analysis
+
+<!--
+    How does this change affect existing ESLint users? Will any behavior
+    change for them? If so, how are you going to minimize the disruption
+    to existing users?
+-->
+
+This RFC describes a set of breaking changes that can be released in the next major version of ESLint. Both ESLint plugin authors and ESLint users will be affected and will be unable to use offending rules until they are updated. This will mostly affect a small number of older ESLint plugins that haven't been updated nor maintained in years.
+
+I did an analysis of the ESLint plugin ecosystem by writing a tool called [eslint-ecosystem-analyzer](https://github.com/bmish/eslint-ecosystem-analyzer) to find out how much of the ecosystem will be affected by these changes.
+
+For the top 100 ESLint plugins, 1% of total rules have options but are missing a schema (or said another way, 5% of rules with options are missing a schema). The percentages are only slightly higher (worse) when the top 1,000 ESLint plugins are considered (the top 1,000 results include many repositories that haven't been maintained nor used in years).
+
+For the top 100 ESLint plugins, 94% of rules are object rules. The percentage is only slightly lower (worse) when the top 1,000 ESLint plugins are considered.
+
+These results indicate that the vast majority of the ecosystem is already ready for these changes and will not be affected.
+
+While the current numbers are already supportive of making these changes, we have a number of means available to further prepare plugins for these changes, which we could optionally employ:
+
+- Communicating about these upcoming changes in blog posts leading up to the major release
```

Comment:
Would updates to RuleTester help?

Reply:
One of the items in this list is:

> Adding deprecation notices inside ESLint in a minor version release, showing a warning when executing linting or running tests for an offending rule

So by that I mean the deprecation notice would show when running rule tests. Does that cover what you are suggesting here?

Reply:
I have added some more details about the deprecation notices when running tests or executing linting for an offending lint rule.

I have also started a branch to demonstrate the deprecation warnings: https://github.com/bmish/eslint/tree/RFC-85-deprecation-warning

Reply:
I don’t think deprecation noticed targeted at end users are a useful approach. How many will actually hunt down the author of a rule (that was maybe included in a shareable config) to let them know?

It seems like targeted messaging in RuleTester of some sort would reach the correct audience.

Reply:
Yep, addressed as part of the other comment thread: https://github.com/eslint/rfcs/pull/85#discussion_r777130444

Reply:
Doh sorry, missed that. Just digging out of my holiday notifications!